tests/framework/e2e/etcdctl.go

```diff
@@ -545,9 +554,73 @@ func (ctl *EtcdctlV3) spawnJsonCmd(output interface{}, args ...string) error {
 	if err != nil {
 		return err
 	}
+	defer cmd.Close()
 	line, err := cmd.Expect("header")
 	if err != nil {
 		return err
 	}
 	return json.Unmarshal([]byte(line), output)
 }
+
+func (ctl *EtcdctlV3) Watch(ctx context.Context, key string, opts config.WatchOptions) clientv3.WatchChan {
+	args := ctl.cmdArgs()
+	args = append(args, "watch", key)
+	if opts.RangeEnd != "" {
+		args = append(args, opts.RangeEnd)
+	}
+	args = append(args, "-w", "json")
+	if opts.Prefix {
+		args = append(args, "--prefix")
+	}
+	if opts.Revision != 0 {
+		args = append(args, "--rev", fmt.Sprint(opts.Revision))
+	}
+	proc, err := SpawnCmd(args, nil)
+	if err != nil {
+		return nil
+	}
+
+	ch := make(chan clientv3.WatchResponse)
+	curModRev := int64(0)
+	go func() {
+		defer proc.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				close(ch)
+				return
+			default:
+				lines := proc.Lines()
+				for _, line := range lines {
+					var resp clientv3.WatchResponse
+					json.Unmarshal([]byte(line), &resp)
+					if resp.Canceled {
+						close(ch)
+						return
+					}
+					if len(resp.Events) > 0 {
+						// if the mod revision for the event is smaller than the current mod revision,
```

Comment:
This looks suspicious. Why this is a problem for e2e tests and not integration tests? That would mean that etcd client library handles it but etcdctl not?

Reply:
updated to comments. please take a look.